[mpqsolver] Introduce MPQSolver (#10194)
Samsung/ONE: compiler/circle-mpqsolver/README.md

# circle-mpqsolver

> _circle-mpqsolver_ provides light-weight methods for finding a high-quality mixed-precision model within a reasonable time. Currently only _bisection_ method is implemented. 

Reviewer: ```suggestion
_circle-mpqsolver_ provides light-weight methods for finding a high-quality mixed-precision model within a reasonable time.
```

Reviewer: Let's reduce job when introducing a new method.

Author: ok.

## bisection

> So assuming the model is parameterized by single parameter (let's call it _depth_), we can split the model so that all layers with _depth_ parameter below _specified_depth_ will be in the _input_group_, while those layers with _depth_ above _specified_depth_ will be in the _output_group_ (node's _depth_ mimics distance from _input_ to the node).

Reviewer 2: please split lines to come in 80 or 100 columns for better editing.

Author: ok.

# circle-mpqsolver
_circle-mpqsolver_ provides light-weight methods for finding a high-quality mixed-precision model 
within a reasonable time.

> ## bisection

Reviewer: ```suggestion
## Methods

### Bisection
```

So assuming the model is parameterized by single parameter (let's call it _depth_), we can split 
the model so that all layers with _depth_ parameter below _specified_depth_ will be in the 
_input_group_, while those layers with _depth_ above _specified_depth_ will be in the _output_group_ 
(node's _depth_ mimics distance from _input_ to the node). Then _bisection_ searches iteratively 
such a _depth_ parameter which achieves exactly the _target_loss_. In case _input_group_ is 
quantized into Q16 the pseudocode is the following: 
```
    until |_depth_max_ - _depth_min_| <=1 do
        _current_depth_ = 0.5 * (_depth_max_ + _depth_min_)
        if Loss(_current_depth_) < _target_loss_
            _depth_max_ = _current_depth_
        else
            _depth_min_ = _current_depth_
```
, where _Loss(current_depth)_ - measure of discrepancy between input float model and quantized at 
_current_depth_ model. As every iteration halves the range containing _optimal_depth_ it converges 
in a ~_ln(depth_max - depth_min)_ number of iterations although it may produce not the _fastest_ 

> model.

Reviewer: I rewrote the explanation. Please correct if there is something to change.

BTW, how can we compute depth if there are multiple inputs?

---

A model is split into two parts: `front` and `back`. One of them is quantized in uint8 and another in int16. The precision of `front` and `back` is determined by our proxy metric, upperbound of total layer errors.
(See https://github.com/Samsung/ONE/pull/10170#discussion_r1042246598 for more details)

 The boundary between the `front` and the `back` is decided by the _depth_ of operators (_depth_: distance from input to the operator), i.e., given a depth _d_, layers with a depth less than _d_ are included in `front`, and the rest are included in `back`. _Bisection_ performs binary search to find a proper _depth_ which achieves a qerror less than _target_qerror_.

In case _front_ is quantized into Q16 the pseudocode is the following: 

```
    until |_depth_max_ - _depth_min_| <=1 do
        _current_depth_ = 0.5 * (_depth_max_ + _depth_min_)
        if Loss(_current_depth_) < _target_loss_
            _depth_max_ = _current_depth_
        else
            _depth_min_ = _current_depth_
```

, where _Loss(current_depth)_ is the qerror of the mixied-precision model split at _current_depth_. As every iteration halves the remaining range (|_depth_max_ - _depth_min_|), it converges in ~_log2(max_depth)_  iterations.

Author: Thank you. I'll change the description.

Author: > BTW, how can we compute depth if there are multiple inputs?

For this case depth will be maximal distance from input nodes.

## Usage 
Run _circle-mpqsolver_ with the following arguments.  

--data: .h5 file with test data

--input_model: Input float model initialized with min-max (recorded model)

--output_model: Output qunatized mode

> --qerror_ratio: Admissible quantization error ratio (> 0 and  < 1)

Reviewer: ```suggestion
--qerror_ratio: Target quantization error ratio. It should be in [0, 1]. 0 indicates qerror of full uint8 model, 1 indicates qerror of full int16 model.
```

# circle-mpqsolver
_circle-mpqsolver_ provides light-weight methods for finding a high-quality mixed-precision model 
within a reasonable time.

## Methods

### Bisection
A model is split into two parts: front and back. One of them is quantized in uint8 and another in 
int16. The precision of front and back is determined by our proxy metric, upperbound of total layer 

> errors. (See #10170 (comment) for more details)

Reviewer 2: > #10170 (comment)

This does not show to a link and `comment isn't point to which comment to the issue.
It would be better to (1) fix this to a real link so that reader can jump or (2) insert the comment here and add a reference link to the source.

Author: Ahhh... sorry. I'll fix it.
